In [1]:
import pandas as pd
import numpy as np

import re
from pymongo import MongoClient
import datetime

from sklearn.model_selection import train_test_split

In [4]:
categories = ['Space', 'Technology', 'Business', 'Science']
categories = ['Business']

In [5]:
client = MongoClient('mongodb://tweedyRead:102030Aa@18.206.103.114:27017/TWEEDY')
db=client['TWEEDY']

In [6]:
cat_list = []                          # a list to collect seperate dataframes of each category

for i, cat in enumerate(categories):

    df_cat = pd.DataFrame(db.Tweet.find({"edInput"  : { "$in": [ 1,2 ] }, 
                               "topicName": cat, 
                               "crDate"   : {"$gt": datetime.datetime(2020, 3, 1, 0, 0, 0)}, 
                               "crDate"   : {"$lt": datetime.datetime(2020, 8, 1, 0, 0, 0)} }, 
                              {"_id":False, "text":1, "edInput":1, "topicName":1} ))
    
    df_cat.loc[df_cat['edInput'] ==1, 'edInput'] = i      # Change category (1) to    : i
    df_cat.loc[df_cat['edInput'] ==2, 'edInput'] = i+10   # Change non-category (0) to: i+10
       
    cat_list.insert(i, df_cat)    # Seperate dataframes are collected

In [7]:
df = pd.concat(cat_list)    # I concatenated all dataframes into one

df.reset_index(drop=True, inplace=True)

In [8]:
df.shape

(92750, 3)

In [10]:
pd.set_option('display.max_colwidth', 300)    # To be able to see complete tweets in dataframe

df.sample(1000).sort_values(by=['edInput'])      # Random samples

edInput  \
57421      0.0   
25089      0.0   
8869       0.0   
18179      0.0   
69551      0.0   
28389      0.0   
47053      0.0   
48429      0.0   
13244      0.0   
45057      0.0   
44074      0.0   
23768      0.0   
62119      0.0   
71453      0.0   
63912      0.0   
48997      0.0   
53651      0.0   
23549      0.0   
68642      0.0   
53683      0.0   
20128      0.0   
67623      0.0   
20000      0.0   
15019      0.0   
19249      0.0   
110        0.0   
52039      0.0   
31960      0.0   
53990      0.0   
69143      0.0   
...        ...   
77488     10.0   
87211     10.0   
75956     10.0   
91438     10.0   
81667     10.0   
73002     10.0   
78213     10.0   
78059     10.0   
86416     10.0   
72619     10.0   
80467     10.0   
73818     10.0   
80555     10.0   
86933     10.0   
86655     10.0   
84498     10.0   
91758     10.0   
84186     10.0   
82274     10.0   
90363     10.0   
84125     10.0   
71982     10.0   
79403     10.0   
80665     10.0   
90756     10.0   
79364     10.0   
74575     10.0   
85903     10.0   
84354     10.0   
74373     10.0   

                                                                                                                                                                                                                                                              text  \
57421                                                                                                                                                                  9 free apps college students can use to find a job or an internship https://t.co/d7sna7sOzA   
25089                                                                                        Leaders need to take stock of the unwritten social norms in their organization and make sure they align with the culture they want to create. https://t.co/MwcWlcs0Ai   
8869                                                                                                                                        Your data is too cheap; the Czech Republic shows the way to make tech giants pay https://t.co/jcEY1xzc95 via @bopinion   
18179                         When employees trust their organizations, they’re more productive.\nWhen customers trust businesses, they show them loyalty. \nNeuroscientist @pauljzak breaks down why people trust — and when they don’t.\nhttps://t.co/p6oAAl95b2   
69551                            Stitcher releases its first podcasting report, with 10 years of data trends; a pandemic update shows a dip in listening at commute times and a lunchtime rise (Stitcher Blog)\n\nhttps://t.co/r5sjgVeZQ5\nhttps://t.co/W4qVXdQ8sG   
28389                                                                                                                                         Classic cars can be an expensive investment — this app aims to make investing in them easier https://t.co/OaTQUT6TVf   
47053                                                                                                                                                                                          Chris Hadfield's first blast off into space https://t.co/OKn1tVbwDp   
48429                                                                                                                                                    Electricity data reveals the truth about our lockdown lifestyles @entrylevelrebel https://t.co/UmuDIIvqh7   
13244                                                                                                                                                    Here's why nobody wants to read your content and what to do about it. @amybgeorge https://t.co/3aVDuVRHC2   
45057                                                                                                U.S. Agriculture Secretary Sonny Perdue sees a “less than 10%” chance that U.S. farmers will get another round of trade aid this year https://t.co/He3vE9PlxB   
44074   

In [27]:
for i, cat in enumerate(categories):
    print(i, cat)

0 Space
1 Technology
2 Business
3 Science


In [11]:
df['edInput'].value_counts().sort_index()

0.0     71730
10.0    21020
Name: edInput, dtype: int64

### Category 2  (Business) seems absent. This is a problem but number of Non-Business Tweets (cat.12) is  92,750!!! which makes it the biggest class.

### Category 3 (Science) impose a heavy imbalance to dataset. I think we can reach better results if we remove Category 3 but I will postpone it to further studies for the time being.
### Data is  imbalanced. I will use class_weight='balanced'

In [80]:
df['text'][df['edInput']==12].sample(10)    # samples from non-Business tweets

110910                                     Andrew Yang is right: Wage gains for many Americans have been frustratingly hard to come by. \n\nBut the problem isn't automation, it's fragmentation. The U.S. economy is becoming increasingly geographically segregated https://t.co/mTvcJULoba https://t.co/4iKCyX4i0h
177450                                                                                                                                                                                                        At least 30 people were killed and 38 injured in attacks in Nigeria’s northeast https://t.co/QjltXB2GGH
158448    Assistant Professor Julien Clement studied eSports and found that players who quickly react to a new challenge are more likely to win. These tactics also apply in business: Teams who think like gamers can get the upper hand in rapidly changing industries. https://t.co/4LSmSXXsQT https://t.co/qNw...
122805                                                                

In [40]:
text_train, y_train = df.text, df.edInput

In [41]:
print("type of text_train: {}".format(type(text_train)))
print("\nlength of text_train: {}".format(len(text_train)))
print("\ntext_train[500]:\n{}".format(text_train.iloc[500]))
print("\ntext_train[1000]:\n{}".format(text_train.iloc[1000]))

type of text_train: <class 'pandas.core.series.Series'>

length of text_train: 192082

text_train[500]:
Hubble Telescope Discovers 'Living Fossil' Galaxy in Our Milky Way's Backyard https://t.co/AuvF9Y7htR https://t.co/1oOc7xBrEw

text_train[1000]:
NASA and Navajo Nation Partner in Understanding the Universe https://t.co/5CmwJoi5Gu https://t.co/xdgd5yJPfP


### We have 192,082 Tweets in total

In [1]:
from sklearn.feature_extraction.text import CountVectorizer

In [68]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import SGDClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC

from sklearn.preprocessing import MaxAbsScaler
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.pipeline import Pipeline
from sklearn.pipeline import make_pipeline

In [44]:
print(text_train.iloc[600]) # just as an axample

Earth's Magnetic North Pole Was Moving So Fast, Geophysicists Had to Update the Map https://t.co/T4fwXVdvcn https://t.co/EIkNS4lRpN


In [45]:
# in order to remove URL links, I used below regex.

text_noURL = [re.sub(r"\bhttps://t.co/\w+", ' ', doc) for doc in text_train]
text_noURL = pd.Series(text_noURL)

In [46]:
text_noURL[600]    # I check the same tweet to see URL has been removed

"Earth's Magnetic North Pole Was Moving So Fast, Geophysicists Had to Update the Map    "

In [48]:
text_noURL.sample(5)  #random samples

50517                             How to do Hypothesis Testing : A Beginner Guide For Data Scientist \n\n  v/ @DataScienceL\n#DataScience #BigData #MachineLearning #AI #ArtificalIntelligence\nCc @evankirstel @ahier @SpirosMargaris @JimMarous @andi_staub @schmarzo @KirkDBorne
17808                                                                                                                                                                                                      There’s always an opportunity for a pee joke in space. #Apollo13AtHome  
175665    I honestly don't understand why we have had nearly ten years of pretending that the Adani Carmichael mine is viable. \n\nHere's some really simple maths explaining why it just doesn't stand up financially:\n\n1. Plan is for a 10m ton per year mine, rising to 20mtpa
51168                                                                                                                                                                       

In [51]:
text_noURL = [re.sub("\'ve", " have ", doc) for doc in text_noURL]  # just a simple text cleaning step
text_noURL = pd.Series(text_noURL)

In [56]:
print("type of text_noURL: {}".format(type(text_noURL)))
print("\nlength of text_noURL: {}".format(len(text_noURL)))
print("\ntext_train samples:\n\n{}".format(text_noURL.sample(3)))

type of text_noURL: <class 'pandas.core.series.Series'>

length of text_noURL: 192082

text_train samples:

74625             It's Trump versus the experts: To hold an election during a pandemic, vote-by-mail is the option that will save lives.  
122675                     Facebook warns that it would be a huge win for China if U.S. lawmakers derail theLibra cryptocurrency project  
166596    WATCH: Chains like Starbucks and Pret A Manger are adjusting their stores so they can adjust to social distancing guidelines    
dtype: object


In [57]:
x_train, x_test, y_train, y_test = train_test_split (text_noURL, y_train, test_size=0.2)

In [59]:
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

print("Number of stop words: {}".format(len(ENGLISH_STOP_WORDS)))
print("\nEvery 10th stopword:\n{}".format(list(ENGLISH_STOP_WORDS)[::10]))

Number of stop words: 318

Every 10th stopword:
['often', 'between', 'fifteen', 'nine', 'herself', 'thru', 'former', 'whole', 'nor', 'herein', 'for', 'someone', 'alone', 'empty', 'ie', 'formerly', 'where', 'has', 'anyone', 'not', 'namely', 'my', 'due', 'her', 'we', 'every', 'also', 'though', 'nowhere', 'everything', 'it', 'thick']


In [60]:
vect = CountVectorizer(min_df=3, stop_words="english").fit(x_train)           # min_df=3, stop_words="english"
X_train = vect.transform(x_train)

print("X_train: stop words removed. (min_df=3):\n\n{}".format(repr(X_train)))

X_train: stop words removed. (min_df=3):

<153665x30382 sparse matrix of type '<class 'numpy.int64'>'
	with 1652528 stored elements in Compressed Sparse Row format>


### We have 153,665 x 68,129 matrix in training data; 68,129 of which are tokens which is actually too many but I will leave it as it is. We can limit them manually to 15,000 by max_features later.

In [63]:
feature_names = vect.get_feature_names()

print("Number of features: {}".format(len(feature_names)))
print("\nFirst 50 features:\n{}".format(feature_names[:50]))
print("\nFeatures 10050 to 0100:\n{}".format(feature_names[10050:10100]))
print("\nEvery 1500th feature:\n{}".format(feature_names[::1500]))

Number of features: 30382

First 50 features:
['00', '000', '0000', '000lbs', '000th', '001', '0010', '003', '00am', '00p', '00pm', '01', '0125', '0130', '0151', '017', '01pm', '02', '020', '0200', '02pm', '03', '03am', '03pm', '04', '0450', '04b', '04pm', '05', '054', '05am', '05pm', '06', '06am', '07', '0756', '07pm', '08', '08pm', '09', '092', '0934', '0944', '096', '099', '09pm', '0a', '0b', '10', '100']

Features 10050 to 0100:
['europa', 'europaclipper', 'europarl_en', 'europe', 'european', 'europeancommission', 'europeans', 'euros', 'eurovision', 'eurozone', 'eusmartcities', 'eutelsat', 'ev', 'ev_rat', 'eva', 'evacuate', 'evacuated', 'evacuation', 'evacuations', 'evacuees', 'evade', 'evaded', 'evading', 'evakaili', 'evaluate', 'evaluated', 'evaluates', 'evaluating', 'evaluation', 'evan', 'evanderburg', 'evangelist', 'evanki', 'evankirstel', 'evankohn', 'evans', 'evaporate', 'evaporated', 'evaporating', 'evaporation', 'evas', 'evasion', 'eve', 'evening', 'evenings', 'evenly', 'ev

### I will try only one algorithm with these simple preperation steps just to see if it gives us a promising result.

In [75]:
sss = StratifiedShuffleSplit(n_splits=5, test_size =0.2, random_state=0)

logreg = LogisticRegression(solver='sag', multi_class='auto', max_iter=10000, 
                             class_weight='balanced', random_state=0, n_jobs=-1)

In [77]:
logreg_scores = cross_val_score(logreg, X_train, y_train, cv=sss, n_jobs=-1)

print("Mean cross-validation accuracy (LogReg no URL): {:.3f}".format(np.mean(logreg_scores)))

Mean cross-validation accuracy (LogReg no URL): 0.743


In [110]:
print ("Cross-validation scores are", ", ".join("%.3f" % f for f in logreg_scores))

Cross-validation scores are 0.749, 0.745, 0.747, 0.725, 0.747


### Crossvalidation accuracy is %74.3 and values are very stable.
### Included categories are ['Space', 'Technology', 'Business', 'Science'] .

### This has just been a trial to see if we get a promising result in multiclass classification of four main calsses in WowSo Twitter dataset for further considerations. The dataset is heavily imbalanced but the initial result with only one algorithm is not bad. 

### I think we can reach better results by Random forest, SVM or Boosting algorithms together with TF-IDF, lemmatization and dimensionality reduction by TruncatedSVD in further studies.
